source/connection_manager.c

```diff
@@ -460,16 +487,127 @@ static MQTTStatus_t prvMQTTConnect( bool xCleanSession )
     LogInfo( ( "Session present: %d\n", xSessionPresent ) );
 
     /* Resume a session if desired. */
-    if( ( xResult == MQTTSuccess ) && !xCleanSession )
+    if( ( xResult == MQTTSuccess ) && ( xCleanSession == false ) )
     {
         xResult = MQTTAgent_ResumeSession( &xGlobalMqttAgentContext, xSessionPresent );
+
+        /* Resubscribe to all the subscribed topics. */
+        if( ( xResult == MQTTSuccess ) && ( xSessionPresent == false ) )
+        {
+            xResult = prvHandleResubscribe();
+        }
     }
 
     return xResult;
 }
 
 /*-----------------------------------------------------------*/
 
+static MQTTStatus_t prvHandleResubscribe( void )
+{
+    MQTTStatus_t xResult = MQTTBadParameter;
+    int32_t lIndex = 0;
+    uint16_t usNumSubscriptions = 0U;
+
+    /* These variables need to stay in scope until command completes.
+     * Dynamically allocating memory for these will help to optimize the memory
```

Comment:
Note that using dynamic allocations also increases the likelihood of memory fragmentation, so it's not something you'd always want to do.  I'd simply say that this demo uses static allocations for these because they have to persist during the subscribe operation and delete the _"Dynamically allocating..."_ sentence.